library/alloc/src/alloc/struct_alloc.rs

```diff
@@ -0,0 +1,216 @@
+use crate::alloc::Global;
+use crate::fmt;
+use core::alloc::{AllocError, Allocator, Layout};
+use core::fmt::{Debug, Formatter};
+use core::marker::PhantomData;
+use core::mem;
+use core::ptr::NonNull;
+
+#[cfg(test)]
+mod tests;
+
+/// Allocator that adds appropriate padding for a repr(C) struct.
+///
+/// This allocator takes as type arguments the type of a field `T` and an allocator `A`.
+///
+/// Consider
+///
+/// ```rust,ignore (not real code)
+/// #[repr(C)]
+/// struct Struct {
+///     t: T,
+///     data: Data,
+/// }
+/// ```
+///
+/// where `Data` is a type with layout `layout`.
+///
+/// When this allocator creates an allocation for layout `layout`, the pointer can be
+/// offset by `-offsetof(Struct, data)` and the resulting pointer points is an allocation
+/// of `A` for `Layout::new::<Struct>()`.
+pub(crate) struct StructAlloc<T, A = Global>(A, PhantomData<*const T>);
+
+impl<T, A> StructAlloc<T, A> {
+    #[allow(dead_code)]
+    /// Creates a new allocator.
+    pub(crate) fn new(allocator: A) -> Self {
+        Self(allocator, PhantomData)
+    }
+
+    /// Computes the layout of `Struct`.
+    fn struct_layout(data_layout: Layout) -> Result<Layout, AllocError> {
```

Comment:
Do you have a specific reason, why you dont' use `Layout::extend`? See [`Affix::allocation_layout`](https://docs.rs/alloc-compose/0.5.0/src/alloc_compose/affix.rs.html#252)

Reply:
At the time I wasn't aware of this function. Either way, this function is written to optimize performance given that we know that the layout of `T` will be known at compile time.

Reply:
The compiler is able to optimize out `Layout::extend`, when `T` is known at compile time: https://godbolt.org/z/1YP8f6

Reply:
Compare https://godbolt.org/z/Mon6j1

Reply:
In the `Rc` use case, `data_layout` is also known at compile time. Anyway, the allocator may be used in other places as well, so it would be better to use `Layout::extend` and optimize that function instead. I don't see a point in duplicating code logic, when there is a function for that exact use case.

Reply:
>In the Rc use case, data_layout is also known at compile time.

It is not. data_layout is the layout passed to the allocator.

>so it would be better to use Layout::extend and optimize that function instead

I do not know if it is possible in general to optimize this function. There are different formulas to calculate padding etc., each optimized for different cases. If the compiler were able to tell the optimizer that the alignment is always a power of two and > 0, then those formulas might be optimized identically. But the compiler does not do this at this time. Note that in my optimized function I use two different formulas to minimize the number of runtime-only values.

Reply:
> It is not. data_layout is the layout passed to the allocator.

Yes, but the allocator is only called behind `Rc`, and the type of `T` in `Rc<T>` is known. Actually, I just noticed, `StructAlloc::allocate` is never called when constructing `(A)Rc`, which should be changed.

> If the compiler were able to tell the optimizer that the alignment is always a power of two and > 0, then those formulas might be optimized identically.

I opened a PR a while ago: #75281 but it wasn't merged, maybe we should try it again.

Reply:
> Yes, but the allocator is only called behind `Rc`, and the type of `T` in `Rc<T>` is known.

T can be a DST.

Reply:
> I opened a PR a while ago: #75281 but it wasn't merged, maybe we should try it again.

I experimented with this a while ago and LLVM did not make use of such assertions.